apollo-router-core/src/cache.rs

```diff
@@ -0,0 +1,176 @@
+use crate::{CacheResolver, CacheResolverError};
+use futures::lock::Mutex;
+use lru::LruCache;
+use std::cmp::Eq;
+use std::collections::HashMap;
+use std::fmt;
+use std::hash::Hash;
+use tokio::sync::broadcast::{self, Sender};
+
+/// A result from a cache get().
+pub type CacheResult<V> = Result<V, CacheResolverError>;
+
+/// A caching map optimised for slow value resolution.
+///
+/// The CachingMap hold values in an LruCache. Values are loaded into the cache on a cache miss and
+/// the cache relies on the resolver to provide values. There is no way to manually remove, update
+/// or otherwise invalidate a cache value at this time. Values will be evicted from the cache once
+/// the cache_limit is reached.
+pub struct CachingMap<K, V> {
+    cached: Mutex<LruCache<K, CacheResult<V>>>,
+    #[allow(clippy::type_complexity)]
+    wait_map: Mutex<HashMap<K, Sender<(K, CacheResult<V>)>>>,
+    cache_limit: usize,
+    resolver: Box<dyn CacheResolver<K, V> + Send + Sync>,
+}
+
+impl<K, V> fmt::Debug for CachingMap<K, V> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("CachingMap")
+            .field("cache_limit", &self.cache_limit)
+            .finish()
+    }
+}
```

Comment:
Use `derivative` instead

Reply:
Done in commit: https://github.com/apollographql/router/pull/235/commits/64bd3ebb7d2011df4c50a97d6d6fe26dcb06ca47